src/domain/dataverse/aggregate.ts

```diff
@@ -0,0 +1,176 @@
+          ),
+        setHasNext: (hasNext: boolean): void =>
+          set(
+            (state): ImmerSetReturnType => ({
+              query: { ...state.query, hasNext }
+            })
+          ),
+        setError: (error: DataverseQueryError): void =>
+          set(
+            (state): ImmerSetReturnType => ({
+              query: { ...state.query, error }
+            })
+          ),
+        setFilters: (newFilters: DataverseQueryFilters): void => {
+          const { clearDataverse } = get()
+          clearDataverse()
+          return set(
+            (state): ImmerSetReturnType => ({
+              query: { ...state.query, filters: { ...state.query.filters, ...newFilters } }
+            })
+          )
+        },
+        setLanguage: (newLng: string): void => {
+          const {
+            clearDataverse,
+            query: { setError }
+          } = get()
+          if (!newLng.length) {
+            return setError(
+              new Error(
+                `Oops.. An error occurred in <setLanguage> call.. Parameter cannot be empty: <${newLng}>..`
+              )
+            )
+          }
+          clearDataverse()
+          return set(
+            (state): ImmerSetReturnType => ({
+              query: { ...state.query, language: newLng }
+            })
+          )
+        }
+      },
+      clearDataverse: (): void =>
+        set(
+          (): ImmerSetReturnType => ({
+            dataverse: []
+          })
+        ),
+      addToDataverse: (newDataverse: DataverseEntity): void =>
+        set(
+          (state): ImmerSetReturnType => ({
+            dataverse: [...state.dataverse, ...newDataverse]
+          })
+        ),
+      loadDataverse: (): T.Task<void> | void => {
+        const { addToDataverse, query, dataverse } = get()
+        const { setIsLoading, setHasNext, setError, limit, language, filters } = query
+        const offset = dataverse.length
+
+        setIsLoading(true)
+        setError(null)
+
+        const checkLanguageInvariant = (): E.Either<Error, void> =>
+          pipe(
+            E.tryCatch(
+              () => {
+                if (!language.length) {
+                  throw new Error(
+                    'Oops.. An error occurred when trying to load Dataverse.. Language cannot be empty..'
+                  )
+                }
+              },
+              reason =>
+                reason instanceof Error
+                  ? reason
+                  : new Error('Oops.. An unspecified error occurred..')
+            )
+          )
+
+        const fetchDataverse = (): T.Task<void> =>
+          pipe(
+            gateway.retrieveDataverse(language, limit, offset, filters),
+            TE.match(
+              e => {
+                setIsLoading(false)
+                setError(e)
+              },
+              r => {
+                setIsLoading(false)
+                setHasNext(r.query.hasNext)
+                addToDataverse(r.data)
```

Comment:
There's a potential race condition here. If `fetchDataverse` is called twice, it's possible that the same result could be inserted more than once. To be confirmed.

The same for `isLoading`.

Reply:
1. will not solve the race condition, I presume. If you capture the condition in the asynchronous function, then you will have a risk of overwriting with 3 (or more) concurrent calls. Moreover, changing the interface contract makes the caller responsible for deciding which function to call in which case, when it should not.
2. `isLoading` is your flag since it's part of the state. Then check & eject.
3. use a simple lock mechanism. Then check & stack.

up to you.